src/magic_folder/config.py

```diff
@@ -0,0 +1,419 @@
+"""
+Configuration and state database interaction.
+
+See also docs/config.rst
+"""
+
+from os import (
+    mkdir,
+    urandom,
+)
+from base64 import (
+    urlsafe_b64encode,
+)
+
+import attr
+
+import sqlite3
+
+from nacl.signing import (
+    SigningKey,
+)
+from nacl.encoding import (
+    Base32Encoder,
+)
+
+from twisted.internet.endpoints import (
+    serverFromString,
+)
+from twisted.python.filepath import (
+    FilePath,
+)
+
+from .snapshot import (
+    LocalAuthor,
+)
+
+
+class ConfigurationError(Exception):
+    """
+    The configuration is unusable for some reason
+    """
+
+
+_global_config_version = 1
+
+_global_config_schema = """
+CREATE TABLE version
+(
+    version INTEGER  -- contains one row, set to 1
+);
+
+CREATE TABLE magic_folders
+(
+    name          TEXT PRIMARY KEY,  -- UTF8 name of this folder
+    location      TEXT               -- UTF8 path to this folder's configuration/state
+);
+
+CREATE TABLE config
+(
+    api_endpoint TEXT,                -- Twisted server-string for our HTTP API
+    tahoe_client_url TEXT            -- HTTP URL of our Tahoe-LAFS client
+);
+"""
+
+_magicfolder_config_version = 1
+
+_magicfolder_config_schema = """
+CREATE TABLE version
+(
+    version INTEGER  -- contains one row, set to 1
+);
+
+CREATE TABLE stash
+(
+    path TEXT    -- the path to our stash-directory
+);
+
+CREATE TABLE config
+(
+    author_name          TEXT PRIMARY KEY,  -- UTF8 name of the author
+    author_private_key   TEXT,              -- base32 key in UTF8
+    stash_path           TEXT,              -- local path for stash-data
+    collective_dircap    TEXT,              -- read-capability-string
+    upload_dircap        TEXT,              -- write-capability-string
+    magic_directory      TEXT,              -- local path of sync'd directory
+    poll_interval        INTEGER            -- seconds
+);
+
+CREATE TABLE local_snapshots
+(
+    id VARCHAR(256) PRIMARY KEY, -- identifier (hash of .. stuff)
+    name TEXT,                   -- the (mangled) name in UTF8
+    metadata TEXT,               -- arbitrary JSON metadata in UTF8
+    content_path TEXT            -- where the content is sitting (path, UTF8)
+);
+"""
+## XXX "parents_local" should be IDs of other local_snapshots, not
+## sure how to do that w/o docs here
+
+
+def create_global_configuration(basedir, api_endpoint, tahoe_client_url):
+    """
+    Create a new global configuration in `basedir` (which must not yet exist).
+
+    :param FilePath basedir: a non-existant directory
+
+    :param unicode api_endpoint: the Twisted server endpoint string
+        where we will listen for API requests.
+
+    :param unicode tahoe_client_url: the Twisted client endpoint
+        string where we will contact our Tahoe LAFS client WebUI.
+
+    :returns: a GlobalConfigDatabase instance
+    """
+    if basedir.exists():
+        raise ValueError(
+            "'{}' already exists".format(basedir.path)
+        )
+    mkdir(basedir.path)
```

Comment:
Race condition here.  `mkdir` and handle the EEXIST instead?  Also there's `basedir.createDirectory` and `basedir.makedirs`.